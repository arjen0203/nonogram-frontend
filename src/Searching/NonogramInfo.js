import React, {Component} from 'react';

class NonogramInfo extends Component {
    render() {
        return (
            <div className="nonoogram-info" onClick={() => this.props.onClick()}>
                <div className="gram-info-name">{this.props.info.name}</div>
                <div className="gram-info-creater">By: {this.props.info.creater}</div>
            </div>
        );
    }
}

export default NonogramInfo;