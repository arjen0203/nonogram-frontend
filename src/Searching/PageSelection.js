import React, {Component} from 'react';

class PageSelection extends Component {
    render() {
        return (
            <div className="page-nav">
                <button className="page-nav-button" onClick={() => this.props.changePage(0)}>{"<<"}</button>
                <button className="page-nav-button" onClick={() => this.props.changePage(this.props.currentPage - 1)}>{"<"}</button>
                <input className="page-nav-input" type="number" value={this.props.displayNumber} onChange={(e) => this.props.changeDisplayNumber(e.target.value)} onBlur={(e) => this.props.changePage(e.target.value)}></input>
                <button className="page-nav-button" onClick={() => this.props.changePage(this.props.currentPage + 1)}>{">"}</button>
                <button className="page-nav-button" onClick={() => this.props.changePage(this.props.maxPage)}>{">>"}</button>
            </div>
        );
    }
}

export default PageSelection;