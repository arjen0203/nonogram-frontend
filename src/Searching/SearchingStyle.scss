@import './../colors';

.center {
    justify-content: center;
    display: flex;

    .nonogram-selection {
        display: flex;
        justify-content: center;
        flex-direction: column;
        width: 90%;
        align-items: center;
        margin-top: 40px;

        .nonogram-infos {
            display: flex;
            justify-content: space-around;
            margin-top: 20px;
            margin-bottom: 20px;
            flex-wrap: wrap;
            width: 100%;

            .nonoogram-info {
                margin: 1% 2.5% 1% 2.5%;
                display: flex;
                flex-direction: column;
                align-items: center;
                width: 15%;
                background-color: $silverSandLight;
                padding: 2%;
                border-radius: 20px;
                border: 5px solid $silverSand;
                cursor: pointer;
                transition: 0.5s;

                .gram-info-name {
                    font-weight: bold;
                    font-size: 24px;
                }
                .gram-info-creater {
                    font-style: italic;
                    font-size: 18px;
                }

                &:hover {
                    background-color: $silverSand;
                }
            }
        }
        .page-nav {
            background-color: $silverSand;
            padding: 10px;
            border-radius: 8px;

            .page-nav-button {
                display: inline-block;
                margin: 5px;
                padding: 5px;
                width: 26px;
                height: 26px;
                background-color: $blackCoral;
                color: white;
                cursor: pointer;
                text-decoration: none;
                border: none;
                font-weight: bold;
                text-align: center;
                &:hover {
                    background-color: $blackCoralLight;
                }
            }
            input::-webkit-outer-spin-button,
            input::-webkit-inner-spin-button {
                -webkit-appearance: none;
                margin: 0;
            }

            .page-nav-input {
                width: 25px;
                padding: 4px;
                text-align: right;
                text-align: center;


            }
        }
    }
}