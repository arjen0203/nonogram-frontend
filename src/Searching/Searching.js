import React, {Component} from 'react';
import NonogramInfo from "./NonogramInfo";
import './SearchingStyle.scss'
import PageSelection from "./PageSelection";

class Searching extends Component {
    constructor() {
        super();

        const params = new URLSearchParams(document.location.search.substring(1));

        let page = params.get('page');
        if (page === null) page = 0;

        this.state = {
            loading: false,
            currentPage: 0,
            maxPage: 0,
            nonogramInfos: [],
            displayNumber: page
        }
    }

    componentWillUnmount() {
        this._isMounted = false;
    }

    setNonogramInfos(data) {
        let nonogramInfos = [];

        for (let i = 0; i < data.length; i++) {
            nonogramInfos.push({id: data[i].id, name: data[i].name, creater: data[i].createrName});
        }
        this.setState({nonogramInfos});

    }

    getInfoComps() {
        let output = this.state.nonogramInfos.map((info) =>
            <NonogramInfo key={"NI" + info.id} info={info} onClick={() => this.goToNonogram(info.id)}></NonogramInfo>
        );
        return output;
    }

    goToNonogram(id){
        this.props.history.push("/solving/" + id);
    }

    componentDidMount() {
        this.setInfos();
    }

    setInfos(){
        this._isMounted = true;

        const params = new URLSearchParams(document.location.search.substring(1));

        let page = params.get('page') - 1;

        if (page < 1){
            page = 0;
        }

        const URL = 'https://nonograms.nl/api/nonogram/getAll?page=' + page;
        fetch(URL)
            .then( async res => {

                if (res.ok) {
                    const maxPage = await res.headers.get("TotalPages");
                    res.json().then(data => {
                        this.setNonogramInfos(data)
                        this.setState({loading: false, currentPage: page + 1, maxPage: maxPage, displayNumber: page + 1})
                    })
                } else {
                    const error = await res.text();
                    this.setState({error});
                }
            }).catch(() => this.setState({error: "Could not communicate with server"}));


    }

    changePage(page){
        this.props.history.push("/solving?page=" + page)
        this.setInfos();
    }

    changDisplauNumber(displayNumber){
        this.setState({displayNumber})
    }

    render() {
        let info;
        if (!this.state.loading) info = this.getInfoComps();
        return (
            <div data-testid="searching" className="center">
                {this.state.loading ? (
                    <div>{this.state.error}</div>
                ) : (
                    <div className="nonogram-selection">
                        <div className="nonogram-infos">{info}</div>
                        <PageSelection changePage={(page) => this.changePage(page)} displayNumber={this.state.displayNumber} currentPage={this.state.currentPage}
                                       maxPage={this.state.maxPage} changeDisplayNumber={() => this.changDisplauNumber()}></PageSelection>
                    </div>
                )}
            </div>
        );
    }
}

export default Searching;