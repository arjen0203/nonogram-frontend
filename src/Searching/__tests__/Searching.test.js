import React from "react";
import ReactDOM from 'react-dom';
import Searching from "../Searching";
import {render, cleanup} from "@testing-library/react";
import "@testing-library/jest-dom/extend-expect"
import { act as actRenderer, create } from "react-test-renderer";
import { act } from 'react-dom/test-utils';


afterEach(cleanup);

it('Renders without crashing', function () {
	fetch.dontMockOnce();

	const div = document.createElement('div');
	ReactDOM.render(<Searching></Searching>, div);
});

it('Render correctly', async function () {
	const data = [
		{
			id: 1,
			name: "Big puzzle",
			createrName: "Rens"
		},
		{
			id: 2,
			name: "Smoll puzzle",
			createrName: "Freek"
		}
	];

	fetch.mockResponseOnce(JSON.stringify(data));

	let comp;
	await act(async () => {
		comp = render(
			<Searching></Searching>
		).getByTestId("searching");
	});

	expect(comp).toHaveTextContent("By: Rens");
	expect(comp).toHaveTextContent("Smoll puzzle");
	expect(comp).toHaveTextContent("By: Freek");
	expect(comp).toHaveTextContent("Big puzzle");
});

it('Render correctly', async function () {
	const data = [
		{
			id: 1,
			name: "GREAT STUFF",
			createrName: "Sjoerd"
		},
		{
			id: 2,
			name: "AMAZING",
			createrName: "User567"
		}
	];

	fetch.mockResponseOnce(JSON.stringify(data));

	let comp;
	await act(async () => {
		comp = render(
			<Searching></Searching>
		).getByTestId("searching");
	});

	expect(comp).toHaveTextContent("AMAZING");
	expect(comp).toHaveTextContent("By: Sjoerd");
	expect(comp).toHaveTextContent("GREAT STUFF");
	expect(comp).toHaveTextContent("By: User567");
});

it('matches snapshot', async function () {
	const data = [
		{
			id: 1,
			name: "Big puzzle",
			createrName: "Rens"
		},
		{
			id: 2,
			name: "Smoll puzzle",
			createrName: "Freek"
		}
	];

	fetch.mockResponseOnce(JSON.stringify(data));

	let tree;
	await actRenderer(async () => {tree = create(<Searching></Searching>)});
	expect(tree.toJSON()).toMatchSnapshot();
});