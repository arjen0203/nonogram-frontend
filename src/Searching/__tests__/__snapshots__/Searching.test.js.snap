// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<div
  className="center"
  data-testid="searching"
>
  <div
    className="nonogram-selection"
  >
    <div
      className="nonogram-infos"
    >
      <div
        className="nonoogram-info"
        data-testid="nonoInfo"
        onClick={[Function]}
      >
        <div
          className="gram-info-name"
        >
          Big puzzle
        </div>
        <div
          className="gram-info-creater"
        >
          By: 
          Rens
        </div>
      </div>
      <div
        className="nonoogram-info"
        data-testid="nonoInfo"
        onClick={[Function]}
      >
        <div
          className="gram-info-name"
        >
          Smoll puzzle
        </div>
        <div
          className="gram-info-creater"
        >
          By: 
          Freek
        </div>
      </div>
    </div>
    <div
      className="page-nav"
      data-testid="pageSelect"
    >
      <button
        className="page-nav-button"
        onClick={[Function]}
      >
        &lt;&lt;
      </button>
      <button
        className="page-nav-button"
        onClick={[Function]}
      >
        &lt;
      </button>
      <label>
        <input
          className="page-nav-input"
          data-testid="pageInput"
          onBlur={[Function]}
          onChange={[Function]}
          type="number"
          value={1}
        />
      </label>
      <button
        className="page-nav-button"
        onClick={[Function]}
      >
        &gt;
      </button>
      <button
        className="page-nav-button"
        onClick={[Function]}
      >
        &gt;&gt;
      </button>
    </div>
  </div>
</div>
`;
